pkg/server/sotw/v2/server.go

```diff
@@ -218,6 +235,9 @@ func (s *server) process(stream Stream, reqCh <-chan *discovery.DiscoveryRequest
 			if !seen || responseNonce == nonce {
 				if cancel, seen := values.cancellations[typeUrl]; seen && cancel != nil {
 					cancel()
+					if err := processAll(); err != nil {
+						return err
+					}
```

Comment:
If cancel is asynchronous, then the buffer could be full right here and CreateWatch would block. Imagine cache decided to push right before calling `cancel()`, gets `cancel()` but is already on the way to push it, which happens right after `processAll`.

Reply:
hm, I didn't think that cancel can be async

Reply:
Are there any real-world use-cases for async cancel?

Reply:
@kyessenov If we are dealing with async cancel-s/etc - maybe we should actually consider intermediate slice of responses? No size configuration, no need to deal with blocking stuff, can support both ads and non-ads with single code and no deadlocks.
Going the
```CreateWatch(request *Request, responses chan<- Response) (resp Response, cancel func())```
route basically fixes the deadlock problem, still making block on buffer overflow possible.